Implement some of the Transform2D API in rust

setzer22 opened the pull request:
As we discussed, here's an initial implementation of the Transform2D. The implementation is currently untested, so I'm marking it as a work-in-progress. I'm just posting this early so we can discuss if you thing this is the right approach :+1:

The code closely follows the exposed GDScript API and even copies its documentation strings. The Implementation is adapted from Godot's C++ source code, delegating to glam (via Vector2) when necessary.

I decided against adding a conversion between glam::Affine2 and Transform2D because it was not immediately clear to me if the mapping would be 1:1 (godot's Transform2D does not necessarily need to be an affine transformation, although it's assumed in certain places).

Instead, I decided to just follow the C++ implementation, which was simple enough, to ensure results would be predictable when porting code from GDScript. We are still using glam underneath when delegating all of `Vector3`'s operations, which happens in many places.

Bromeon commented on gdnative-core/src/core_types/transform2d.rs:
```diff
@@ -8,3 +8,188 @@ pub struct Transform2D {
+    }
+
+    /// Returns the transform's rotation (in radians).
+    pub fn get_rotation(&self) -> f32 {
```
Getters should not have a `get_` prefix according to current convention. For `scale()` it can be a bit confusing, but we would need to change this everywhere.

Bromeon reviewed (commented):
Thanks a lot!

Style-wise, you could use `Self` to refer to the current type, and use `#[inline]` for methods. I'm not a huge fan of the latter, but we have currently configured Clippy to check it. Might need to do a benchmark whether it has an impact on speed/code size.

One thing I'm not very sure about, is the getters/setters for translation (=origin), rotation and scale. Their presence kind of implies that the transform is affine, but with field public, this invariant cannot be upheld. What do we return if someone creates a non-affine transform?

Also, I would like to go with the existing convention for getters that have no `get_()` prefix.

I haven't checked the algorithms themselves yet, this is where tests might come in handy.

Bromeon commented:
Fixes #290
Fixes #719

jacobsky reviewed (changes requested):
Thank you for the PR! It looks good to me and I agree with all points on @Bromeon 's feedback.

One thing that I did notice is that the clippy lints requesting #[inline] fail, so please make sure to check with clippy that all of the necessary inlining has been done :)

setzer22 commented:
Thanks @Bromeon  and @jacobsky  for the reviews! I think I've addressed all of the comments (please let me know if you think there's anything else missing).

I will have a look now at some tests against the godot API to make sure the current implementation is working as expected.

setzer22 commented:
Just uploaded a first draft for the tests. Since I'm new at this and there's quite a bit more work to test the remaining functions, I'd like to wait for some early feedback before continuing. If you thing the current strategy's fine, I can go ahead and add the remaining methods :+1:

Bromeon commented:
The current MSRV (minimum supported Rust version) 1.48 does not have `clamp` stabilized yet, so we get this error:
```rs
error[E0658]: use of unstable library feature 'clamp'
   --> gdnative-core/src/core_types/transform2d.rs:190:19
    |
190 |         let dot = f32::clamp(v1.dot(v2), -1.0, 1.0);
    |                   ^^^^^^^^^^
    |
    = note: see issue #44095 <https://github.com/rust-lang/rust/issues/44095> for more information
```

If we want to keep that compatibility, we'd probably have to write our own clamp for now...

Let's see if there are other issues:
bors try

Bromeon commented:
bors try

jacobsky commented:
@setzer22 Thank you for the hard work! I think everything looks good to go!

bors r+

jacobsky commented:
bors r+

jacobsky commented:
Odd, I'm listed as a reviewer on this PR, but bors doesn't seem to trust me :rofl:
![image](https://user-images.githubusercontent.com/1459514/139509365-5b16187d-e49b-46c0-a705-a7f49d4c0851.png)

bors r+

jacobsky commented:
bors r+

jacobsky commented:
bors r+
